docs/content/topics/contributor-guide/hacking.md

```diff
@@ -0,0 +1,281 @@
+identified three tools that, combined, have streamlined Brigade development to
+such an extreme extent that they've become part of our recommended development
+environment.
+
+To continue, you will need to install the latest stable versions of:
+
+* [KinD](https://kind.sigs.k8s.io/#installation-and-usage): Runs
+  development-grade Kubernetes clusters in Docker.
+
+* [ctlptl](https://github.com/tilt-dev/ctlptl#how-do-i-install-it): Launches
+  development-grade Kubernetes clusters (in KinD, for instance) that are
+  pre-connected to a local image registry.
+
+* [Tilt](https://docs.tilt.dev/#macoslinux): Builds components from source and
+  deploys them to a development-grade Kubernetes cluster. More importantly, it
+  enables developers to rebuild and replace running components with the click of
+  a button.
+
+Follow the installation instructions for each of the above.
+
+To launch a brand new Kind cluster pre-connected to a local image registry:
+
+```shell
+$ make hack-kind-up
+```
+
+To build and deploy all of Brigade from source:
+
+```shell
+$ tilt up
+```
+
+Tilt will also launch a web-based UI running at
+[http://localhost:10350](http://localhost:10350). Visit this in your web browser
+and you will be able to see the build and deployment status of each Brigade
+component, complete with logs. Once Tilt has all of Brigade up and running, the
+Brigade API server will be exposed (without TLS) on `localhost:31600`, so if you
+wish to log in using the `brig` CLI as the "root" user:
+
+```shell
+$ brig login --server http://localhost:31600 --root
+```
+
+The root user's password is `F00Bar!!!`.
+
+> ⚠️&nbsp;&nbsp;Tilt is often configured to watch files and automatically
+> rebuild and replace running components when their source code is changed. _We
+> have deliberately disabled this._ Each of our components takes long enough to
+> build that we have discovered it's better for our CPUs if things aren't
+> _constantly_ building and rebuilding in the background and build instead only
+> when we choose. The web UI makes it easy to identify components whose source
+> has been altered. They can be rebuilt and replaced with one mouse click.
+
+When you are done with Tilt, interrupt the running `tilt up` process with
+`ctrl + c`. Components _will remain running in the cluster_, but Tilt will no
+longer be in control. If Tilt is restarted later, it will retake control of the
+already-running components.
+
+If you wish to undeploy everything Tilt has deployed for you, use `tilt down`.
+
+To destroy your KinD cluster, use `make hack-kind-down`.
+
+> ⚠️&nbsp;&nbsp;`make hack-kind-down` deliberately leaves your local registry
+> running so that if you resume work later, you are doing so with a local
+> registry that's already primed with most layers of each of Brigade's images.
+> If you wish to destroy the registry, use:
+>
+> ```shell
+> $ docker rm -f brigade-github-gateway-dev-cluster-control-plane
```

Comment:
Should this be `brigade-dev-registry`?

Reply:
It should be, but I also copied/pasted that directly from my terminal, so I may need to see if something isn't working as I had intended.

I thought I had this and all the peripherals set to use the _same_ local registry.

Reply:
Oh... I just literally copied the wrong container name after a `docker ps`. Will fix.

Reply:
Fixed